### PR TITLE
Issue1635 expm frechet

### PR DESCRIPTION
**Pull request by sriharikrishna**
This is an implementation of `scipy.linalg.expm_frechet` to compute the derivatives of matrix exponentiation. It was mentioned in #1635  as a way to provide derivatives for `jax.scipy.linalg.expm` which has itself been implemented in #1940 .This pull request does not set the `defvjp` rules to use `scipy.linalg.expm_frechet` because testing such an enhancement seems nontrivial. **EDIT: The defjvp rule has been added**

There are a few things to note:
1) `scipy.linalg.expm_frechet_algo_64` creates an identity matrix in the following way:`ident = np.identity(n)`. This is problematic because it does not ensure that `ident` and `A` have the same `dtype`. Note further that `scipy.linalg.expm`  does set the `dtype` of the identity matrix to be the same as `A`. The `dtype` of the output of  `scipy.linalg.expm_frechet` can therefore be different from its arguments. For this reason, `_CheckAgainstNumpy` is called with `check_dtypes=False`.

Major thanks to @jhueckelheim for identifying this problem.

2) There was a need to set the `dtype `of the constant 2.0 within `_diff_pade13` to `A.dtype` to ensure numerical correctness. 
Originally the code was:`
A = A * 2.0**-s`
We found that the `dtype` of the expression `A * 2.0**-s` is not necessarily the same as `A`.
The code was changed to:
` two = np.array([2.0],A.dtype)`
`A = A * two[0]**-s`

### CODE REVIEW

File: jax/scipy/linalg.py [hunk 1]
```diff
@@ -379,3 +379,196 @@ def _pade13(A):
...
+    U5, V5, Lu5, Lv5, s5 = _diff_pade5(A, E, ident)
+    U7, V7, Lu7, Lv7, s7 = _diff_pade7(A, E, ident)
+    U9, V9, Lu9, Lv9, s9 = _diff_pade9(A, E, ident)
+    # scaling
+    U13, V13, Lu13, Lv13, s13 = _diff_pade13(A, E, ident, A_norm_1)
+    U = np.select((A_norm_1<=ell_table_61_local),(U3,U5,U7,U9),U13)
+    V = np.select((A_norm_1<=ell_table_61_local),(V3,V5,V7,V9),V13)
+    Lu = np.select((A_norm_1<=ell_table_61_local),(Lu3,Lu5,Lu7,Lu9),Lu13)
+    Lv = np.select((A_norm_1<=ell_table_61_local),(Lv3,Lv5,Lv7,Lv9),Lv13)
+    s = np.select((A_norm_1<=ell_table_61_local),(s3,s5,s7,s9),s13)
```
Comment (shoyer) [inline jax/scipy/linalg.py]: Something to consider here, which is likely a consideration for the `expm` as well, is that this is actually computing all orders of Pade approximations, and then throwing away all but one. (Note that `A_norm_1` is typically a scalar.)

This suggests to me that we should probably be considering one of two approaches here:
1. Use `lax.cond` instead of `np.select`/`np.where`, to avoid computing unneeded terms. This works pretty well on CPUs but unfortunately accelerators are less amenable to control flows.
2. If higher order Pade approximations are strictly better (but more expensive and sometimes unnecessary), then it might be better to _always_ use them, and not bother making the low order approximation.

I don't inherently know which of (1) and (2) is better -- it might depend on the details of the hardware. It might be useful to made some measurements.

File: jax/scipy/linalg.py [hunk 2]
```diff
@@ -379,3 +379,196 @@ def _pade13(A):
...
+        1.17e1,
+        )
+        
+@jit
+def expm_frechet_algo_64(A, E):
+    ident = np.eye(*A.shape, dtype=A.dtype)
+    A_norm_1 = np_linalg.norm(A, 1)
+    """
+    Subset of the Maximal values ell_m of ||2**-s A||
+    such that the backward error bound does not exceed 2**-53.
```
Comment (shoyer) [inline jax/scipy/linalg.py]: I'm guessing that we could use simpler algorithms for float32?

File: jax/scipy/linalg.py (hunk 1 above)
Comment (sriharikrishna) [reply jax/scipy/linalg.py]: These are interesting points to consider!

The theory does not explicitly discuss using a higher Pade approximation. So, I cannot say definitively that using a higher Pade approximation is strictly better.

The original implementation by Al-Mohy in [Matlab,](http://eprints.maths.manchester.ac.uk/1442/3/expm_new.zip) `scipy.sparse.linalg.expm`, `scipy.linalg.expm_frechet` all use the appropriate Pade approximation.  `tf.linalg.expm` computes all the approximations, before selecting the appropriate one. 

Using only pade7 for dtype=float32 and dtype=complex64, and only pade13 or dtype=float64 and dtype=complex128 passes _CheckAgainstNumpy. I am not sure if that is necessarily instructive.

Because tf computes all the approximations, I am not sure that there is a better approach available. I can modify the implementation to use lax.cond as you suggested.

File: jax/scipy/linalg.py (hunk 2 above)
Comment (sriharikrishna) [reply jax/scipy/linalg.py]: Possibly. It could be argued that a different table should be used for float32 which could lead to the computation of lower Pade approximations. But I don't know what that table should be. 

If the question is whether Squaring and Scaling is the appropriate algorithm for float32, I don't know. S&S seems to be accepted widely. I have not examined Ward1977 or Sidje1998.

File: jax/scipy/linalg.py (hunk 1 above)
Comment (sriharikrishna) [reply jax/scipy/linalg.py]: There are two possible implementations with `lax.cond`.
1. Recursive version that requires `_diff_pade13` to be called each time because of type mismatch in the value of `s` returned by `_diff_pade3`,..`_diff_pade9` and `_diff_pade13`. I have used this approach.
2. Non-recursive version that requires only the appropriate Pade approximation to be executed. This option is, however, not jittable because of the boolean `and` in  `lax.cond`'s `pred` argument.
```
    args = (A, E, ident, A_norm_1)
    U, V, Lu, Lv, s = _diff_pade3(args)
    res = (U, V, Lu, Lv, s)
    
    U, V, Lu, Lv, s = lax.cond(ell_table_61[3]<A_norm_1 and A_norm_1<=ell_table_61[5],
          args, lambda args: _diff_pade5(args),
          res, lambda res: res)
    
    U, V, Lu, Lv, s = lax.cond(ell_table_61[5]<A_norm_1 and A_norm_1<=ell_table_61[7],
          args, lambda args: _diff_pade7(args),
          res, lambda res: res)
        
    U, V, Lu, Lv, s = lax.cond(ell_table_61[7]<A_norm_1 and A_norm_1<=ell_table_61[9],
          args, lambda args: _diff_pade9(args),
          res, lambda res: res)
    
    U, V, Lu, Lv, s = lax.cond(ell_table_61[9]<A_norm_1 ,
          args, lambda args: _diff_pade13(args),
          res, lambda res: res)
```

Comment (jekbradbury) [reply jax/scipy/linalg.py]: Does it work to use `&` instead of `and`?

Comment (sriharikrishna) [reply jax/scipy/linalg.py]: Ah yes. Replacing `&` by `and` is jittable. Thank you for suggesting it. 

The non-recursive version however also requires a `select` statement for `_diff_pade13(args)` because the following code throws a type mismatch exception for the argument s (the code in the comment above was incorrect):
```
    U, V, Lu, Lv, s = lax.cond(ell_table_61[9]<A_norm_1 ,
          args, lambda args: _diff_pade13(args),
          res, lambda res: res)
```
This means the that the non-recursive version always computes `_diff_pade3` and `_diff_pade13` whereas the recursive version always computes one of `_diff_pade3`, `_diff_padet5`, and_`diff_pade9` as well as `_diff_pade13`. The non-recusive version however performs more conditional expressions. So the recursive version is ultimately preferable.

**Comment by sriharikrishna** [comment]
@shoyer I have not been able to get GPU timing measurements comparing the use of `lax.cond` to `np.select/np.where`. This is because I have not yet been successful in installing jaxlib with cuda enabled.  

In any case, I have used `lax.cond` for now.  What do you think should be the next steps?

**Comment by sriharikrishna** [comment]
I am going to close this PR for the moment till I am able to validate the defjvp rule.

**Comment by sriharikrishna** [comment]
I have fixed the defjvp rule. The forward mode derivatives are correct and pass test_util.check_jvm(). I am facing an obscure error when I test the reverse mode derivatives:
'''
2020-03-06 13:47:58.806380: F external/org_tensorflow/tensorflow/compiler/xla/service/hlo_instructions.h:962] Check failed: ShapeUtil::GetLeafCount(shape()) == parameter_replicated_at_leaf_buffers.size() (2 vs. 3)
Abort trap: 6
'''
I understand that the defjvm convenience function is supposed to automatically set the defvjm rule as well. 

Any help in solving the error would be appreciated.

**Comment by shoyer** [comment]
That's definitely a weird error/bug!

I don't know what's going on there, but I don't think we can expect defining a JVP be to sufficient to get us backward mode (VJP) here. In order for that to, the JVP rule needs to be transposable, but we use `fori_loop` in the JVP calculation (which isn't transposable).

Instead, I think we want `expm_frechet` to be a JAX primitive. Differentiation is a linear transformation, so the JVP rule should call `expm_fretchet` inside of itself, but we will need to define a transpose rule to get backward mode working. I suspect that's pretty straightforward, though it's not immediately obvious to me what it should be yet.

That said, I would suggest not worrying about auto-diff rules in this PR, and just getting in a basic working version of expm_fretchet first. We can save the auto-diff for later.

**Comment by sriharikrishna** [comment]
Okay. I  consider the code in the PR a basic working version of expm_fretchet. It passes `_CheckAgainstNumpy` and `_CompileAndCheck`. Should I expect it to be merged and consider replacing the fori_loop as a separate enhancement to support auto-diff rules?

**Comment by terhorst** [comment]
@shoyer A project I'm working on is blocking on this issue (autodiff for `expm()`), so following your suggestion I took a look at defining `expm_frechet()` as a primitive. @gnecula's nice tutorial made that pretty easy, but I got stuck when it came to defining the XLA translation rule. It seems like you'd have to implement the whole XLA computation graph embodied by `expm_frechet()` by hand. Is there some way to tell JAX to just use what's already implemented for this part?

**Comment by hawkinsp** [comment]
@terhorst I think you're looking to use `xla.lower_fun` in the implementation of your translation rule.

**Comment by terhorst** [comment]
@hawkinsp Thank you, that's exactly what I was looking for. I declared the translation via 
```
jax.xla.translations[expm_frechet_p] = jax.xla.lower_fun(expm_frechet_impl)
```
But when I do this, I get an error which, essentially, says that the output of `expm_frechet_impl` gets unpacked along the leading/batched axis into an XLA Tuple. Later functions which are expecting an array of dimension `[batch, n, n]` instead get a `batch`-dimensional tuple of `n x n` arrays. Am I somehow using this function incorrectly? It seems consistent with other usages I found in the source.

**Comment by hawkinsp** [comment]
@terhorst I can't say for sure without seeing your code, but note the function passed to `xla.lower_fun` should produce a tuple (*), so perhaps try something like `return (out,)`?

(*) I just added a `multiple_results` keyword argument to `lower_fun` at head, so you could also try syncing up to head and using `xla.lower_fun(..., multiple_results=False)`.

**Comment by NeilGirdhar** [comment]
I'm just a casual observer here, but I thought `custom_transforms` was deprecated?  Shouldn't this use `custom_jvp` instead?

**Comment by shoyer** [comment]
I think this PR predates custom_jvp/custom_vjp.

**Comment by sriharikrishna** [comment]
I will amend the PR to use custom_jvp/custom_vjp.

**Comment by hawkinsp** [comment]
Any chance we could check this in, even using the deprecated `custom_transforms`? We can check it in as is and update it in a subsequent PR.

If you are short on time I'm happy to update the PR and check it in for you.

**Comment by shoyer** [comment]
+1 for checking in incremental progress here.

**Comment by hawkinsp** [comment]
I rebased this PR on top of head, and updated it to use the new `custom_jvp` API.
